README.md
# Mara Markdown Documentation

Display markdown documentation in mara UI

[![Build Status](https://travis-ci.org/jankatins/mara-markdown-docs.svg?branch=master)](https://travis-ci.org/jankatins/mara-markdown-docs)
[![PyPI - License](https://img.shields.io/pypi/l/mara-markdown-docs.svg)](https://github.com/jankatins/mara-markdown-docs/blob/master/LICENSE)
[![PyPI version](https://badge.fury.io/py/mara-markdown-docs.svg)](https://badge.fury.io/py/mara-markdown-docs)
[![Slack Status](https://img.shields.io/badge/slack-join_chat-white.svg?logo=slack&style=social)](https://communityinviter.com/apps/mara-users/public-invite)


This package displays configured documentation in markdown format in the UI:

- Convert markdown to html via [markdown-it](https://github.com/markdown-it/markdown-it)
- Supports a folder structure (single level) in the menu
- Supports [mermaid](https://mermaid-js.github.io/mermaid/#/) diagrams
- Serves referenced images (`.png`, `.jpg`, `.gif`) in markdown and `.txt` files (for example config, etc) 
  if these are in the same folder as the markdown file

&nbsp;

## Screenshot

![Screenshot](mara-markdown-docs_800px.png) 

&nbsp;

## Installation

To use the library directly, use pip:

```
python3 -m pip install mara-markdown-docs

# or directly from git
python3 -m pip install git+https://github.com/mara/mara-markdown-docs.git
```

&nbsp;

## Configuration

Assuming you configure via `app/local_setup.py`:

```python
"""Configures the docs functionality"""

import pathlib
from mara_app.monkey_patch import patch

import mara_markdown_docs.config

@patch(mara_markdown_docs.config.documentation)
def documentation() -> dict:
    """Dict with name -> path to markdown file.

    If name contains a single '/' it will be shown in a submenu. Multiple '/' are not allowed.
    The insertion order is mostly preserved (folders are grouped in the menu)."""

    repo_root_dir = pathlib.Path(__file__).parent.parent

    # Cases matter in path!
    docs = {
        'Pipeline/Marketing': repo_root_dir / 'app/pipelines/marketing/README.md',
        'Developer/Setup': repo_root_dir / 'README.md',
        'Developer/Code Conventions': repo_root_dir / 'code_conventions.md',
    }

    return docs
```


CHANGELOG.md
# Changelog

## 0.9.0 (2020-09-29)

* Serve images and `.txt` files from the same folder as the markdown document
* Upload to pypi

## 0.1.0 (2020-08-22)

- Move to Github


tests/test_mara_markdown_docs.py
import mara_markdown_docs.config
import mara_app.app
import mara_page.acl
import pathlib
from mara_app.monkey_patch import patch


import pytest


# Workaround for https://github.com/mara/mara-page/pull/6
@patch(mara_page.acl.current_user_has_permissions)
def fixed_current_user_has_permissions(resources):
    return list(map(lambda resource: [resource, True], resources))

repo_root_dir = pathlib.Path(__file__).parent.parent
repo_readme = repo_root_dir / 'README.md'

def _client():

    @patch(mara_markdown_docs.config.documentation)
    def documentation() -> dict:
        """This is a test docstring"""


        # Cases matter in path!
        docs = {
            'xyz_Marketing_xyz': repo_root_dir / 'app/pipelines/marketing/README.md',
            'Developer/Setup': repo_readme,
            'Developer/Code Conventions': repo_root_dir / 'code_conventions.md',
            'test': repo_root_dir / 'tests/docs/test.md'
        }

        return docs

    flask_app = mara_app.app.MaraApp()

    with flask_app.test_client() as client:
        flask_app.testing = True
        yield client

client = pytest.fixture(_client)

def test_redirect_to_full_url(client):
    r = client.get('/docs/developer')
    print(r.data)
    assert b'/docs/developer/' in r.data, 'no redirect'
    assert b'Redirect' in r.data, 'no redirect'
    assert r.status_code == 308, 'No redirect: status'


def test_docs_in_menu(client):
    r = client.get('/mara-app/navigation-bar')
    print(r.data)
    assert b'>Developer</' in r.data, 'Missing folder structure'
    assert b'xyz_Marketing_xyz' in r.data, 'missing entry'
    assert b'href="/docs/developer/code_conventions/"' in r.data, 'missing URL'
    assert b'href="/mara-app/configuration#mara_markdown_docs.config"' in r.data, 'missing entry for config'


def test_docs_in_config(client):
    r = client.get('/mara-app/configuration')
    print(r.data)
    assert b'id="mara_markdown_docs.config"' in r.data, 'Missing config section'
    assert str(repo_readme).encode('utf8') in r.data, 'missing repo readme entry'

def test_doc(client):
    r = client.get('/docs/developer/setup/')
    print(r.data)
    assert b'mara-markdown-docs.git' in r.data, 'Missing link to repo'

def test_overview(client):
    r = client.get('/docs/')
    print(r.data)
    assert b'xyz_Marketing_xyz' in r.data, 'missing entry'
    assert b'href="/docs/developer/code_conventions/"' in r.data, 'missing URL'


def test_404_directory(client):
    r = client.get('/docs/developer/')
    print(r.data)
    assert b'404' in r.data, 'Missing page is not missing'
    assert r.status_code == 404, 'Missing page is not missing: status'


def test_404_page(client):
    r = client.get('/docs/developer/missing/')
    print(r.data)
    assert b'404' in r.data, 'Missing page is not missing'
    assert r.status_code == 404, 'Missing page is not missing: status'

def test_docs_file_endpoint(client):
    r = client.get('/docs/test/test.txt')
    print(r.data)
    assert b'$Test doc$' in r.data
    assert r.status_code == 200

def test_docs_file_endpoint_404(client):
    r = client.get('/docs/test/test2.txt')
    print(r.data)
    assert r.status_code == 404

def test_docs_file_endpoint_dot_dot(client):
    # flask prevents / in the last segment, so this also guards against ..
    r = client.get('/docs/test/../docs/test.txt')
    print(r.data)
    assert r.status_code == 404

def test_docs_file_endpoint_subfolder(client):
    # flask prevents / in the last segment, so also no subfolder
    r = client.get('/docs/test/docs/test.txt')
    print(r.data)
    assert r.status_code == 404

def test_docs_file_endpoint_not_allowed_filetype(client):
    r = client.get('/docs/test/secrets.conf')
    print(r.data)
    assert b'$secrets$' not in r.data
    assert r.status_code == 405

@pytest.fixture
def client_unauth():
    import mara_page.acl
    import mara_markdown_docs.docs
    orig_func = mara_page.acl.current_user_has_permissions

    @patch(mara_page.acl.current_user_has_permissions)
    def current_user_has_permissions(resources):
        def _exclude_docs(resource):
            if resource is mara_markdown_docs.docs.documentation_acl_resource:
                return [resource, False]
            return [resource, True]
        return list(map(_exclude_docs, resources))

    yield from _client()

    patch(mara_page.acl.current_user_has_permissions)(orig_func)

def test_unauth_docs(client_unauth):
    r = client_unauth.get('/docs/developer/setup')
    assert r.status_code == 403, 'Forbidden page is not forbidden'
    r = client_unauth.get('/docs/')
    assert r.status_code == 403, 'Forbidden page is not forbidden'
    r = client_unauth.get('/mara-app/configuration')
    assert r.status_code == 200, 'config should still be allowed'



mara_markdown_docs/mermaid_md_plugin.py
from markdown_it import MarkdownIt

from markdown_it.common.utils import unescapeAll, escapeHtml
from markdown_it.token import Token

from typing import List
import re

# orig JS: token.info.match(/^graph (?:TB|BT|RL|LR|TD);?$/)
GRAPH_RE = re.compile(r'^graph (TB|BT|RL|LR|TD)?;?$')
CHART_TYPES = ('mermaid', 'gantt', 'sequenceDiagram', 'classDiagram', 'gitGraph')



def mermaid_plugin(md: MarkdownIt):
    """Render fenced code blocks for mermaid diagrams

    Example:

    ```mermaid
    graph LR
      A --- B
    ```

    The plugin does not add the javascript to actually render the mermaid diagram, just wraps the
    diagram in a '<div class="mermaid">...</div>' as needed by mermaid. Add something like the following
    to your page to let mermaid render the diagram:

    ```JS
    <script src="mermaid.js"></script>

    <script type="text/javascript">
    window.addEventListener('load', function(event) {
       mermaid.initialize({ /* mermaid config */ });
    });
    </script>

    See https://mermaid-js.github.io/mermaid/#/usage for more details

    """

    # save it here to play nice with other plugins which also overwrite/extend this rule
    orig_fence = md.renderer.rules['fence']

    # in add_render_rule(), self will be bound to the md.renderer instance
    def mermaid_fence(self, tokens: List[Token], idx: int, options, env):
        token = tokens[idx]
        info = token.info or ''

        if not (info in CHART_TYPES or GRAPH_RE.match(info)):
            return orig_fence(tokens, idx, options, env)

        # If language is a chart type, we do not need to render that but if it is already part of the
        # graph definition, we have to add it
        if info in CHART_TYPES:
            content = token.content.strip()
        else:
            content = info + '\n' + token.content.strip()

        # Also allow \n as an alternative to </br> in labels (needs enablement in the mermaid config)
        content = content.replace('\\n','<br/>')
        content = escapeHtml(content)
        return f'<div class="mermaid">{content}</div>'

    md.add_render_rule('fence', mermaid_fence)


mara_markdown_docs/static/markdown-it-naive-mermaid.js
// A really stupid renderer for code blocks to a mermaid div

const NaiveMermaidPlugIn = (md, opts)=> {
  const defaultRenderer = md.renderer.rules.fence.bind(md.renderer.rules);

  md.renderer.rules.fence=(tokens, idx, opts, env, self)=>{
    const token = tokens[idx];
    var x;
    if (token.info === 'mermaid' || token.info === 'gantt' || token.info === 'sequenceDiagram' || token.info === 'classDiagram' || token.info === 'gitGraph') {
       x = `${token.content.trim()}`;
    } else {
       x = `${token.info} \n ${token.content.trim()}`;
    }
    // also recognice \n as a linebreak
    const code = x.replace(/\\n/g,'<br/>');
    if (token.info === 'mermaid' || token.info === 'gantt' || token.info === 'sequenceDiagram' || token.info === 'classDiagram' || token.info === 'gitGraph' || token.info.match(/^graph (?:TB|BT|RL|LR|TD);?$/)) {
        return `<div class="mermaid">${code}</div>`
    }
    return defaultRenderer(tokens, idx, opts, env, self)
  }
}



mara_markdown_docs/docs.py
import flask
import pathlib
import dataclasses

from mara_page import navigation, response, _, bootstrap, acl

from .config import documentation

docs = flask.Blueprint('docs', __name__, url_prefix='/docs', static_folder='static')

documentation_acl_resource = acl.AclResource(name='Documentation')


@dataclasses.dataclass
class Doc:
    full_doc_id: str
    full_name: str
    path: pathlib.Path

    @property
    def doc_name(self):
        """The Page title which should show up in the menu"""
        doc_name = self.full_name
        if '/' in self.full_doc_id and '/' in self.full_name:
            _, doc_name = self.full_name.split('/', maxsplit=1)
        return doc_name

    @property
    def folder_name(self):
        """The Name of the Folder this doc is in (or empty for root)"""
        folder_name = ''
        if '/' in self.full_doc_id and '/' in self.full_name:
            folder_name, _ = self.full_name.split('/', maxsplit=1)
        return folder_name

    @property
    def ids(self):
        """The folder_id, doc_id of this doc"""
        folder_id = ''
        doc_id = self.full_doc_id
        if '/' in doc_id:
            folder_id, doc_id = doc_id.split('/', maxsplit=1)
        return folder_id, doc_id


def all_docs():
    all_docs = {}
    for full_name, path in documentation().items():
        if full_name.count('/') > 1:
            raise ValueError(f"Only one folder level allowed: {full_name}")
        full_doc_id = f"{full_name.lower().replace(' ', '_')}"
        all_docs[full_doc_id] = Doc(full_doc_id, full_name, path)
    return all_docs


def documentation_navigation_entry():
    children = []
    folders = {}
    for full_doc_id, doc in all_docs().items():
        where_to_append = children
        folder_id = None
        folder_id, doc_id = doc.ids
        if folder_id:
            folder = folders.get(folder_id)
            if not folder:
                folder = navigation.NavigationEntry(label=doc.folder_name, icon='book',
                                                    description=f"Documentation Folder {doc.folder_name}")
                children.append(folder)
                folders[folder_id] = folder
            where_to_append = folder
        child = navigation.NavigationEntry(
            label=doc.doc_name, icon='file', description=f'Documentation for {doc.full_name}',
            uri_fn=lambda _doc_id=doc_id, _folder_id=folder_id: flask.url_for('docs.document',
                                                                              doc_id=_doc_id,
                                                                              folder_id=_folder_id))
        where_to_append.append(child) if isinstance(where_to_append, list) else where_to_append.add_child(child)
    return navigation.NavigationEntry(
        label='Documentation', icon='book', rank=100,
        description='Documentation',
        uri_fn=lambda: flask.url_for('docs.start_page'),
        children=children)


@docs.route('')
@docs.route('/')
@acl.require_permission(documentation_acl_resource)
def start_page():
    links = []
    for full_doc_id, doc in all_docs().items():
        folder_id, doc_id = doc.ids
        links.append(_.li[_.a(href=flask.url_for('docs.document', doc_id=doc_id, folder_id=folder_id))[doc.full_name]])

    return response.Response(title='Docs', html=[
        bootstrap.card(header_left='Table of Content', body=[_.ul[links]])
    ])


@docs.route('/<doc_id>/<file_name>')
@docs.route('/<folder_id>/<doc_id>/<file_name>')
@acl.require_permission(documentation_acl_resource)
def image(doc_id, folder_id="", file_name=""):
    allowed_file_extensions = [
        '.txt',  # e.g. example conf files
        '.png', '.gif', '.jpg', '.jpeg', # images
    ]


    docs = all_docs()
    if folder_id:
        full_doc_id = folder_id + '/' + doc_id
    else:
        full_doc_id = doc_id
    if full_doc_id not in docs:
        raise flask.abort(404, f"Documentation {doc_id} is not known.")
    doc = docs[full_doc_id]
    if not doc.path.exists():
        raise flask.abort(404, f"Documentation {doc_id} is not found ({doc.path}).")

    # make sure we only send out data which is ok...

    # only certain file types
    file = doc.path.parent / file_name
    if file.suffix not in allowed_file_extensions:
        raise flask.abort(405, f"Filetype of {doc_id} / {file_name} is not allowed")

    # flask itself prevents subfolders as the defined routes above do not allow
    # for multiple segments in the filename path
    # # prevent files from subfolders and therefore also '..'
    # if '/' in file_name:
    #     raise flask.abort(405, f"Only filenames allowed.")

    # error on non-existing files
    if not file.exists():
        raise flask.abort(404, f"File {doc_id} / {file_name} is not found ({file}).")

    return flask.send_file(filename_or_fp=file)


@docs.route('/<doc_id>/')
@docs.route('/<folder_id>/<doc_id>/')
@acl.require_permission(documentation_acl_resource)
def document(doc_id, folder_id=""):
    docs = all_docs()
    if folder_id:
        full_doc_id = folder_id + '/' + doc_id
    else:
        full_doc_id = doc_id
    if full_doc_id not in docs:
        raise flask.abort(404, f"Documentation {doc_id} is not known.")
    doc = docs[full_doc_id]
    if not doc.path.exists():
        raise flask.abort(404, f"Documentation {doc_id} is not found ({doc.path}).")
    with doc.path.open() as f:
        md_content = f.read()
        md_escaped = flask.escape(md_content)
        return response.Response(
            title=f'Doc "{doc.full_name}"',
            html=[bootstrap.card(
                body=[
                    _.div(style="display:none")[_.pre(id_='markdown-source')[md_escaped]],
                    _.div(id_='markdown-rendered-content')[
                        _.span(class_='fa fa-spinner fa-spin')[' ']
                    ],
                ]),
                _.script(type="text/javascript")[__render_code],
            ],
            js_files=[
                'https://cdnjs.cloudflare.com/ajax/libs/mermaid/8.7.0/mermaid.min.js',
                'https://cdnjs.cloudflare.com/ajax/libs/markdown-it/11.0.0/markdown-it.min.js',
                'https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.1.2/highlight.min.js',
                flask.url_for('docs.static', filename='markdown-it-naive-mermaid.js'),
            ],
            css_files=[
                'https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.1.2/styles/default.min.css',
            ]
        )


# table render trick from https://github.com/markdown-it/markdown-it/issues/117#issuecomment-109386469

# language=JavaScript
__render_code = """
window.addEventListener('load', function(event) {
var md = window.markdownit({
  html: true,
  linkify: true,
  typographer: true,
  highlight: function (str, lang) {
    if (lang && hljs.getLanguage(lang)) {
      try {
        return hljs.highlight(lang, str).value;
      } catch (__) {}
    }

    return ''; // use external default escaping
  }
}).use(NaiveMermaidPlugIn, {});
// style tables with bootstrap
md.renderer.rules.table_open = function(tokens, idx) {
      return '<table class="table">';
};

var md_text = document.getElementById("markdown-source").innerText;
var md_div = document.getElementById("markdown-rendered-content");
md_div.innerHTML = md.render(md_text);
// using a Mermaid.render(...) in the plugin always resulted in truncated labels for boxes, but doing it after the
// MD is rendered works
mermaid.initialize({
  // allow html tags in text
  securityLevel: 'loose',
  flowchart:{
            // don't go over the surrounding container
            useMaxWidth:true,
            // allow html in labels
            htmlLabels:true
        }
  });
});
"""
#
